### PR TITLE
PXC 8.4.0 refresh - Q2 2024

### PR DESCRIPTION
**Pull request by kamil-holubicki**

### CODE REVIEW

File: mysql-test/mysql-test-run.pl [hunk 1]
```diff
@@ -5356,13 +5352,8 @@ ($)
                            tmpdir              => $opt_tmpdir,
                            user                => $opt_user,
                            vardir              => $opt_vardir,
-<<<<<<< HEAD
                            worker              => $tinfo->{worker} ||
                                                     DEFAULT_WORKER_ID
-||||||| merged common ancestors
-=======
-                           bind_local          => $opt_bind_local
->>>>>>> Percona-Server-8.4.0-1
```
Comment (venkatesh-prasad-v) [inline mysql-test/mysql-test-run.pl]: Any reason why `bind_local` is not needed in PXC?

File: mysql-test/include/mtr_check.sql [hunk 2]
```diff
@@ -117,16 +117,11 @@ BEGIN
...
-<<<<<<< HEAD
                                 'wsrep_start_position',
                                 'auto_increment_offset',
                                 'auto_increment_increment',
                                 'wsrep_data_home_dir',
                                 'keyring_file_data',
-||||||| merged common ancestors
-                                'keyring_file_data',
-=======
->>>>>>> Percona-Server-8.4.0-1
```
Comment (venkatesh-prasad-v) [inline mysql-test/include/mtr_check.sql]: Don't we need to remove `keyring_file_data` here?

File: mysql-test/collections/disabled.def [hunk 3]
```diff
@@ -93,29 +86,16 @@ innodb_gis.gis_split_nan                          : BUG#3129 PXC requires 16k In
 # keyring tests
 keyring_encrypted_file.keyring_migration_bugs @darwin : Bug#31041633
 
-||||||| merged common ancestors
-# keyring tests
-keyring_encrypted_file.keyring_migration_bugs @darwin : Bug#31041633
-
-=======
->>>>>>> Percona-Server-8.4.0-1
```
Comment (venkatesh-prasad-v) [inline mysql-test/collections/disabled.def]: May be `keyring_encrypted_file.keyring_migration_bugs` needs to be enabled as well?

File: mysql-test/collections/disabled.def [hunk 4]
```diff
@@ -93,29 +86,16 @@ innodb_gis.gis_split_nan                          : BUG#3129 PXC requires 16k In
...
 main.ds_mrr-big @solaris                          : BUG#14168107 Test leads to timeout on Solaris on slow sparc servers.
-main.gtid_next_xa_binlog_off : BUG#33650776 Failure of XA COMMIT of prepared txn, can result in txn rollback
-||||||| merged common ancestors
-main.ds_mrr-big @solaris : BUG#14168107 Test leads to timeout on Solaris on slow sparc servers.
-main.gtid_next_xa_binlog_off : BUG#33650776 Failure of XA COMMIT of prepared txn, can result in txn rollback
-=======
-main.ds_mrr-big @solaris : BUG#14168107 Test leads to timeout on Solaris on slow sparc servers.
->>>>>>> Percona-Server-8.4.0-1
-main.print_stacktrace : Bug#36027494 Add mtr test for my_print_stacktrace
+main.gtid_next_xa_binlog_off                      : BUG#33650776 Failure of XA COMMIT of prepared txn, can result in txn rollback
```
Comment (venkatesh-prasad-v) [inline mysql-test/collections/disabled.def]: `main.gtid_next_xa_binlog_off ` needs to be enabled as well

File: scripts/systemd/mysqld_pre_systemd.in [hunk 5]
```diff
@@ -224,7 +224,7 @@ install_db () {
     # mysql_upgrade_history file should be owned by mysql user since MySQL 8.4 (new file in 8.4)
     fix_mysql_upgrade_history
 
->>>>>>> Percona-Server-8.4.0-1
+>>>>>>> ps/release-8.4.0-1
```
Comment (venkatesh-prasad-v) [inline scripts/systemd/mysqld_pre_systemd.in]: Merge conflicts are not resolved in this script. Please resolve them.

File: mysql-test/suite/galera/galera_1node.cnf [hunk 6]
```diff
@@ -31,11 +31,11 @@ innodb_autoinc_lock_mode=2
...
 
-early_plugin_load=keyring_file.so
-keyring_file_data=@ENV.MYSQL_TMP_DIR/mysqld.1/keyring.1
+#loose_early_plugin_load=keyring_file.so
+#loose_keyring_file_data=@ENV.MYSQL_TMP_DIR/mysqld.1/keyring.1
 
-[sst]
-encrypt=4
+#[sst]
+#encrypt=4
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/galera_1node.cnf]: May be we also need to delete the `mysql-test/suite/galera/combinations` file?

File: mysql-test/suite/rpl/t/rpl_drop_binlog_ddl_skip_rewrite.test [hunk 7]
```diff
@@ -1,289 +1,18 @@
...
---echo Executed query: $query
---echo Logged query: $result
-
-# Assert that the query is logged with comments
---assert (`SELECT '$result' = '$query'`)
-
-# Verify that the value of @@binlog_ddl_skip_rewrite is also sent with the DROP TABLE t1 query.
---let $binlog_position=
---let $mysqlbinlog_pipe= | $GREP 'DROP .* t1' -B1 | $GREP 'binlog_ddl_skip_rewrite'
---source include/mysqlbinlog.inc
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_drop_binlog_ddl_skip_rewrite.test]: The newly written test file does not test this functionality.

File: mysql-test/suite/rpl/t/rpl_drop_binlog_ddl_skip_rewrite.test [hunk 8]
```diff
@@ -1,289 +1,18 @@
...
---echo Output: $result
-
-# Assert that value of @@binlog_ddl_skip_rewrite is also sent with the query.
---assert (`SELECT '$result' = "/*!80026 SET @@session.binlog_ddl_skip_rewrite=1*//*!*/;"`)
-
---echo
---echo #
---echo # Test-4: Dropping tables from the connection which has
---echo #         binlog_ddl_skip_rewrite=OFF
---echo #
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/rpl/t/rpl_drop_binlog_ddl_skip_rewrite.test]: The newly written test file does not test this functionality.

File: mysql-test/suite/perfschema/t/dd_version_check.test [hunk 9]
```diff
@@ -13,26 +13,16 @@ drop procedure if exists test.pfs_check_proc;
...
 SELECT SUBSTRING_INDEX(SUBSTRING(properties, LOCATE('PS_VERSION', properties), 30), ';', 1)
-  into @actual_dd_version_tag
   FROM mysql.dd_properties;
 SET SESSION debug= '-d,skip_dd_table_access_check';
 
-SELECT @actual_dd_version_tag as "DATA DICTIONARY TAG";
-
-SELECT SUBSTRING(@actual_dd_version_tag, 12) into @actual_dd_version;
-SELECT @actual_dd_version as "DATA DICTIONARY VERSION";
-
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/perfschema/t/dd_version_check.test]: Most of the parts of this test (including some comments) differs from the test in upstream. Why this change was needed?

File: sql/mysqld.cc [hunk 10]
```diff
@@ -15982,7 +15982,13 @@ PSI_stage_info *wsrep_server_stages[] = {
...
     &stage_wsrep_in_rollback_thread,
     &stage_wsrep_aborter_idle,
-    &stage_wsrep_aborter_active};
+    &stage_wsrep_aborter_active,
+
+    // nbo
+    &stage_wsrep_applied_nbo_writeset,
+    &stage_wsrep_applying_nbo_writeset,
+    &stage_wsrep_nbo_committed,
+    &stage_wsrep_nbo_committing };
```
Comment (venkatesh-prasad-v) [inline sql/mysqld.cc]: Do we also need to backport this change to 8.0 as well?

File: mysql-test/suite/galera/t/suite-setup.sh [hunk 11]
```diff
@@ -0,0 +1,10 @@
+datadir=$1
+name=$2
+cnf_file_name=component_keyring_file.cnf
+
+# create local keyring config in datadir
+echo "{\"path\": \"${MYSQL_TMP_DIR}/${name}/component_keyring_file\", \"read_only\": false }" > ${datadir}/$cnf_file_name
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/suite-setup.sh]: Won't this make all nodes in the cluster use the same `keyring_data_file` in `${MYSQL_TMP_DIR}/${name}/component_keyring_file`? 

Wouldn't it be better if we have it stored in the datadir itself? Something like
```
# create local keyring config in datadir
echo "{\"path\": \"${datadir}/component_keyring_file\", \"read_only\": false }" > ${datadir}/$cnf_file_name
```

File: mysql-test/suite/galera-x/t/prep_stmt_sundries.test [hunk 12]
```diff
@@ -480,20 +480,30 @@ source ../include/wait_for_status_variables.inc;
 --let $assert_cond= [SHOW GLOBAL STATUS LIKE "Prepared_stmt_count", Value, 1] = 0
 --source include/assert.inc
 
+# When Percona Telemetry is built-in it installs the component, which changes conters values
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera-x/t/prep_stmt_sundries.test]: s/conters/counter/

File: mysql-test/suite/galera/t/galera_mysqlpump.test [hunk 13]
```diff
@@ -1,60 +0,0 @@
-#
-# Test the mysqlpump utility with Galera
-#
-
---source include/galera_cluster.inc
```
Comment (venkatesh-prasad-v) [inline mysql-test/suite/galera/t/galera_mysqlpump.test]: Why was this test removed?

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 14]
```diff
@@ -845,22 +829,8 @@ read_cnf()
     # (These files will not be REMOVED on the joiner node)
     cpat=$(parse_cnf sst cpat '.*\.pem$\|.*init\.ok$\|.*galera\.cache$\|.*sst_in_progress$\|.*sst-xb-tmpdir$\|.*gvwstate\.dat$\|.*grastate\.dat$\|.*\.err$\|.*\.log$\|.*RPM_UPGRADE_MARKER$\|.*RPM_UPGRADE_HISTORY$\|.*component_keyring_.*\.cnf$\|.*mysqld.my$')
 
-    # Keep the donor's keyring file
-    cpat+="\|.*/${XB_DONOR_KEYRING_FILE}$"
-
     # Keep the KEYRING_FILE_DIR if it is a subdir of the datadir
     # Normalize the datadir for comparison
-    local readonly DATA_TEMP=$(dirname "$DATA/xxx")
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: This comment also needs to be removed

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 15]
```diff
@@ -1694,6 +1653,9 @@ function initialize_pxb_commands()
         disver="--no-version-check"
     fi
 
+    # KH: TODO: Just for tests, as we don't have PXB 8.4 yet. Remove it
+    disver+=" --no-server-version-check"
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: PXB-8.4 is already released. Does still comment still hold good?

File: scripts/wsrep_sst_xtrabackup-v2.sh [hunk 16]
```diff
@@ -1827,9 +1789,10 @@ if [[ -z $MYSQL_VERSION ]]; then
 fi
 
 # Verify PXB versions we have
-verify_pxb_version "${XTRABACKUP_THIS_VER_PATH}" "${XB_THIS_REQUIRED_VERSION}"
-verify_pxb_version "${XTRABACKUP_PREV_VER_PATH}" "${XB_PREV_REQUIRED_VERSION}"
-verify_pxb_version "${XTRABACKUP_PREV_LTS_VER_PATH}" "${XB_PREV_LTS_REQUIRED_VERSION}"
+# KH: temporary
+#verify_pxb_version "${XTRABACKUP_THIS_VER_PATH}" "${XB_THIS_REQUIRED_VERSION}"
+#verify_pxb_version "${XTRABACKUP_PREV_VER_PATH}" "${XB_PREV_REQUIRED_VERSION}"
+#verify_pxb_version "${XTRABACKUP_PREV_LTS_VER_PATH}" "${XB_PREV_LTS_REQUIRED_VERSION}"
```
Comment (venkatesh-prasad-v) [inline scripts/wsrep_sst_xtrabackup-v2.sh]: PXB-8.4 is already released. Does still comment still hold good?

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]
Hi Kamil! Nice work!!

PR looks good overall and I have a few suggestions/comments. Please find them below.

File: mysql-test/suite/galera/galera_1node.cnf (hunk 6 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/galera_1node.cnf]: No. We need enc_on and enc_off combinations.

File: mysql-test/suite/rpl/t/rpl_drop_binlog_ddl_skip_rewrite.test (hunk 7 above)
Comment (kamil-holubicki) [reply mysql-test/suite/rpl/t/rpl_drop_binlog_ddl_skip_rewrite.test]: I created rpl_drop_binlog_ddl_skip_rewrite_2.test for PXC specific changes

File: mysql-test/mysql-test-run.pl (hunk 1 above)
Comment (kamil-holubicki) [reply mysql-test/mysql-test-run.pl]: done

File: mysql-test/include/mtr_check.sql (hunk 2 above)
Comment (kamil-holubicki) [reply mysql-test/include/mtr_check.sql]: It didn't do any harm. Removed.

File: scripts/systemd/mysqld_pre_systemd.in (hunk 5 above)
Comment (kamil-holubicki) [reply scripts/systemd/mysqld_pre_systemd.in]: Done. However there is one interesting thing I spotted out. We don't have `fix_htp_settings` and `fix_flush_caches` functions which are present in PS

Comment (venkatesh-prasad-v) [reply scripts/systemd/mysqld_pre_systemd.in]: Should we port them to PXC?

File: sql/mysqld.cc (hunk 10 above)
Comment (kamil-holubicki) [reply sql/mysqld.cc]: It is not critical. Was needed because of new test introduced in 8.4 which tested it.

File: mysql-test/suite/perfschema/t/dd_version_check.test (hunk 9 above)
Comment (kamil-holubicki) [reply mysql-test/suite/perfschema/t/dd_version_check.test]: I simply updated the existing test.
As I see the upstream changed the structure of it however.
Aligned to the upstream.

File: mysql-test/suite/galera/t/suite-setup.sh (hunk 11 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/suite-setup.sh]: No. the path is `${MYSQL_TMP_DIR}/${name}`, so it will go to `var/tmp/mysqld.1`, `var/tmp/mysqld.2` and so on.

File: mysql-test/suite/galera-x/t/prep_stmt_sundries.test (hunk 12 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera-x/t/prep_stmt_sundries.test]: fixed

File: scripts/wsrep_sst_xtrabackup-v2.sh (hunk 14 above)
Comment (kamil-holubicki) [reply scripts/wsrep_sst_xtrabackup-v2.sh]: done

File: scripts/wsrep_sst_xtrabackup-v2.sh (hunk 15 above)
Comment (kamil-holubicki) [reply scripts/wsrep_sst_xtrabackup-v2.sh]: done. Leftover from the time when pxb 8.4 was not ready yet

File: scripts/wsrep_sst_xtrabackup-v2.sh (hunk 16 above)
Comment (kamil-holubicki) [reply scripts/wsrep_sst_xtrabackup-v2.sh]: done. Leftover from the time when pxb 8.4 was not ready yet

File: mysql-test/suite/galera/t/galera_mysqlpump.test (hunk 13 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/galera_mysqlpump.test]: mysqlpump was removed in 8.4

File: mysql-test/suite/galera/t/suite-setup.sh (hunk 11 above)
Comment (venkatesh-prasad-v) [reply mysql-test/suite/galera/t/suite-setup.sh]: Ok. Got it.

From the below lines, I thought $name referred to test name. But in fact it is defined as `my $name = $mysqld->name();`
```
7457     if ($mysqld->option('#!run-suite-setup-sh') and $tinfo->{suite_setup_sh} and                                                                                                                              
7458     |   run_sh_script("$tinfo->{suite_setup_sh} $datadir $name")) {                                                                                                                                           
7459     | $tinfo->{'comment'} = "Failed to execute '$tinfo->{suite_setup_sh}'";                                                                                                                                   
7460     | return 1;                                                                                                                                                                                               
7461     }
```

File: mysql-test/collections/disabled.def (hunk 3 above)
Comment (kamil-holubicki) [reply mysql-test/collections/disabled.def]: done. Test was removed by upstream

File: mysql-test/collections/disabled.def (hunk 4 above)
Comment (kamil-holubicki) [reply mysql-test/collections/disabled.def]: test enabled

File: scripts/systemd/mysqld_pre_systemd.in (hunk 5 above)
Comment (kamil-holubicki) [reply scripts/systemd/mysqld_pre_systemd.in]: Not needed

File: scripts/systemd/mysqld_pre_systemd.in [hunk 17]
```diff
@@ -263,24 +100,14 @@ install_db () {
...
-		     --datadir="$datadir" --user=@MYSQLD_USER@ --init-file="$initfile"
+                     --datadir="$datadir" --user=@MYSQLD_USER@ --init-file="$initfile"
     rm -f "$initfile"
 
-<<<<<<< HEAD
     # Generate certs if needed
     if [ -x @bindir@/mysql_ssl_rsa_setup -a ! -e "${datadir}/server-key.pem" ] ; then
         @bindir@/mysql_ssl_rsa_setup --datadir="$datadir" --uid=@MYSQLD_USER@ >/dev/null 2>&1
     fi
     return 0
```
Comment (venkatesh-prasad-v) [inline scripts/systemd/mysqld_pre_systemd.in]: mysql_ssl_rsa_setup has been removed by upstream by commit b75bcb61737b. So there is no need of this block. Perhaps a `exit 0` would be enough.

**Review by venkatesh-prasad-v — CHANGES_REQUESTED** [review]

File: scripts/systemd/mysqld_pre_systemd.in (hunk 17 above)
Comment (kamil-holubicki) [reply scripts/systemd/mysqld_pre_systemd.in]: fixed

**Review by venkatesh-prasad-v — APPROVED** [review]
Nice work on this! Approved.
